File: typo-dsl-scala/src/scala/typo/scaladsl/RuntimeExtensions.scala
package typo.scaladsl

import typo.runtime.*

import scala.jdk.CollectionConverters.*
import scala.jdk.OptionConverters.*

implicit class PgTypeOps[A](private val pgType: PgType[A]) extends AnyVal {
  def nullable: PgType[Option[A]] =
    pgType.opt.to(Bijections.optionalToOption[A])
}

implicit class MariaTypeOps[A](private val mariaType: MariaType[A]) extends AnyVal {
  def nullable: MariaType[Option[A]] =
    mariaType.opt.to(Bijections.optionalToOption[A])
}

implicit class DbTypeOps[A](private val dbType: DbType[A]) extends AnyVal {
  def nullable: DbType[Option[A]] =
    dbType.opt.to(Bijections.optionalToOption[A])
}

implicit class EitherOps[L, R](private val either: Either[L, R]) extends AnyVal {
  def rightOrNone: Option[R] = {
    either.asOptional().toScala
  }

  def leftOrNone: Option[L] = {
    either match {
      case left: typo.runtime.Either.Left[L, R] => Some(left.value())
      case _                                    => None
    }
  }
}

implicit class ArrOps[A](private val arr: typo.data.Arr[A]) extends AnyVal {
  def reshapeOrNone(newDims: Int*): Option[typo.data.Arr[A]] = {
    arr.reshape(newDims*).toScala
  }

  def getOrNone(indices: Int*): Option[A] = {
    arr.get(indices*).toScala
  }
}

implicit class RangeOps[T <: Comparable[T]](private val range: typo.data.Range[T]) extends AnyVal {
  def finiteOrNone: Option[typo.data.RangeFinite[T]] = {
    range.finite().toScala
  }
}

implicit class FragmentBuilderOps(private val builder: Fragment.Builder) extends AnyVal {
  def paramNullable[T](pgType: PgType[T], value: Option[T]): Fragment.Builder = {
    builder.param(pgType.opt(), value.toJava)
  }
}

implicit class FragmentOps(private val fragment: Fragment) extends AnyVal {
  def query[Out](parser: typo.scaladsl.ResultSetParser[Out]): Operation[Out] = {
    fragment.query(parser.underlying)
  }
}

implicit class OperationOptionalOps[T](private val operation: Operation[java.util.Optional[T]]) extends AnyVal {
  def runUncheckedOrNone(c: java.sql.Connection): Option[T] = {
    operation.runUnchecked(c).toScala
  }
}

// Extension for converting Scala Iterator to Java Iterator for streaming inserts
implicit class ScalaIteratorOps[T](private val iterator: Iterator[T]) extends AnyVal {
  def toJavaIterator: java.util.Iterator[T] = iterator.asJava
}

def buildFragment(block: Fragment.Builder => Unit): Fragment = {
  val builder = Fragment.Builder()
  block(builder)
  builder.done()
}

File: foundations-jdbc-dsl-kotlin/src/kotlin/dev/typr/foundations/kotlin/RuntimeExtensions.kt
package dev.typr.foundations.kotlin

import dev.typr.foundations.*
import java.util.Optional

/**
 * Kotlin extension methods for typr-runtime-java that provide:
 * - Nullable types instead of Optional
 * - Kotlin-friendly lambda syntax
 * - Better generic type handling
 */

// ================================
// DbType Extensions
// ================================

/**
 * Kotlin-friendly nullable version of DbType.opt().
 *
 * Java's DbType.opt() returns DbType<Optional<A>>, but Kotlin code should work with A? instead.
 * This extension wraps a DbType<A> to create a DbType<A?> that converts between Optional and nullable
 * at read/write boundaries.
 *
 * Usage:
 *   val nullableText: DbType<String?> = PgTypes.text.nullable()
 *
 * Instead of:
 *   val optionalText: DbType<Optional<String>> = PgTypes.text.opt()  // Java-style
 */
fun <A : Any> DbType<A>.nullable(): DbType<A?> =
    this.opt().to(optionalToNullable())

// ================================
// Either Extensions (value-level)
// ================================

/**
 * Convert Either<L, R> to R? (nullable right value).
 * Uses .orNull() from OptionalExtensions.
 */
fun <L, R> Either<L, R>.rightOrNull(): R? {
    return this.asOptional().orNull()
}

/**
 * Get left value or null.
 */
fun <L, R> Either<L, R>.leftOrNull(): L? {
    return when (this) {
        is dev.typr.foundations.Either.Left -> this.value()
        else -> null
    }
}

// ================================
// Arr Extensions (value-level)
// ================================

/**
 * Reshape array to new dimensions or return null.
 * Uses .orNull() from OptionalExtensions.
 */
fun <A> Arr<A>.reshapeOrNull(vararg newDims: Int): Arr<A>? {
    return this.reshape(*newDims).orNull()
}

/**
 * Get array element at indices or return null.
 * Uses .orNull() from OptionalExtensions.
 */
fun <A> dev.typr.foundations.data.Arr<A>.getOrNull(vararg indices: Int): A? {
    return this.get(*indices).orNull()
}

// ================================
// Range Extensions (value-level)
// ================================

/**
 * Get finite range or null.
 * Uses .orNull() from OptionalExtensions.
 */
fun <T : Comparable<T>> Range<T>.finiteOrNull(): RangeFinite<T>? {
    return this.finite().orNull()
}

// ================================
// Fragment Extensions
// ================================

/**
 * Build a Fragment using Kotlin DSL.
 */
inline fun buildFragment(block: dev.typr.foundations.Fragment.Builder.() -> Unit): Fragment {
    val builder = dev.typr.foundations.Fragment.Builder()
    builder.block()
    return Fragment(builder.done())
}

/**
 * Append a nullable parameter to fragment builder.
 * Converts Kotlin's T? to Java's Optional<T> automatically.
 */
fun <T> dev.typr.foundations.Fragment.Builder.paramNullable(type: DbType<T>, value: T?): dev.typr.foundations.Fragment.Builder {
    return this.param(type.opt(), value.toOptional())
}

/**
 * Kotlin-friendly query method that accepts Kotlin ResultSetParser.
 * Converts Kotlin ResultSetParser to Java ResultSetParser automatically.
 */
fun <Out> dev.typr.foundations.Fragment.query(parser: ResultSetParser<Out>): dev.typr.foundations.Operation<Out> {
    return this.query(parser.underlying)
}

// ================================
// Operation Extensions
// ================================

/**
 * Extension to convert Operation<Optional<T>> results to T? automatically.
 * This handles the common case where Java methods return Optional but Kotlin code expects nullable types.
 */
fun <T> dev.typr.foundations.Operation<java.util.Optional<T>>.runUncheckedOrNull(c: java.sql.Connection): T? {
    return this.runUnchecked(c).orNull()
}
